### DISCUSSION TITLE
file corruption

### DISCUSSION THREAD

**Discussion by trinhpxl** [keiwando/toonsquid#99] [Q&A]
hello ! i decided to make a 3 min animation and my file crashed multiple times D: it started having pink rectangles in replace of my frames, i couldnt back it up either bc the file was just too big ;-; would i email it to you to possibly fix that or D: ? i dont think i can share the project file bc its literally just too big. i even tried google drive and its not working either D:
![2246B8A2-0F7A-45B3-9E19-5BA40EACE666](https://user-images.githubusercontent.com/115050959/208229535-18001f2b-a729-483c-9f45-f338e9d8f12e.png)
![A9043D85-1293-4F58-A89B-DA490F48F8EE](https://user-images.githubusercontent.com/115050959/208229534-6f64c8fe-a728-4aa1-93ad-c4035ee9582d.png)

**Comment by keiwando** [comment]
Hi! Yes, if you could email me the file I can take a look what might have happened there. Is your iPad running low on storage? And did the backup fail or did you skip it?

**Reply by trinhpxl** [reply]
yes it kinda is ! i cant email it bc it exceeds the limit, and i tried google drive but its like, not moving at all in uploading,,, but yea my ipad was low on storage, i bought more icloud storage but i dont think it did anything ?

**Reply by trinhpxl** [reply]
the backup also failed, ot just kept crashing when it would begin

**Reply by keiwando** [reply]
Yes, iCloud storage won't make a difference there. ToonSquid needs available storage space locally on your iPad in order to save your work and for various other tasks. If the storage is not available then the operating system won't let the app save anything and there's unfortunately not really anything it can do about that. How much storage does your iPad say you have left out of how much in total?

I can see if I can maybe detect if your iPad is running too low on storage and show a big warning or something in that case.

**Reply by trinhpxl** [reply]
ahh i see then, 
![309C5FBE-CE97-46A4-A91C-58ACC2D7F949](https://user-images.githubusercontent.com/115050959/208231780-7ff62fd5-0228-4fc2-b821-ff2ac17d7ee2.jpeg)
here ya go :0

**Reply by trinhpxl** [reply]
more info on the proj if ya need (still waiting for the drive thing to upload, its moved a bit)
![65B53F7F-505B-4BB6-96E3-E017608552A0](https://user-images.githubusercontent.com/115050959/208231821-e5498bbb-ec64-461f-b93b-c79877fb6503.png)

**Reply by keiwando** [reply]
Well this looks like you still have plenty of space available. If iOS is showing you the correct numbers here then this should not be a storage space problem.

**Reply by trinhpxl** [reply]
uh oh lmao. is this a bug of some sort then ? 0_0

**Reply by trinhpxl** [reply]
andddd google drive wont work at all either. i guess toonsquid is simply too powerful for its own good /joking

**Reply by keiwando** [reply]
I don't know yet. The backup failing would indicate a lack of storage because it doesn't really do anything special in that case. Is the backup folder empty?

**Reply by trinhpxl** [reply]
no i have 3 backups but they’re like 2 hours old at max

**Reply by keiwando** [reply]
So 3 backups of this project? Can you import the latest one and see how many of the drawings are in there?

**Reply by trinhpxl** [reply]
i tried importing butttt 
![C642AC1C-824F-4182-8D37-B1AC9CA3C033](https://user-images.githubusercontent.com/115050959/208232141-540bf00a-a631-42b7-b98a-ee47058a59e1.png)

**Reply by trinhpxl** [reply]
i do remember though, i imported a video for reference of what i was animating and it spanned to 2k frames - no idea if thats a clue for anything but figured i should say

**Reply by keiwando** [reply]
You have exported a copy of the latest version of the project and saved it somewhere else in order to upload it to Google Drive, right? (Not the backup but the one with the pink squares) If so, can you try to delete it from the project library and then try to import the backup?

**Reply by trinhpxl** [reply]
just exported a copy, saved it to my files app, and rn im uploading it to google drive ! but i did delete the project (like you said), tried importing the backup and it stills says unsupported/corrupted error

**Reply by trinhpxl** [reply]
nvm it just appeared suddenly (the backup imported)

**Reply by keiwando** [reply]
Okay, thanks for all of the info! I will take a look a look at the project file once it has finished uploading to Google Drive. How long would you guess is the upload going to take? (e.g. ~10min, ~1hour, multiple hours)?

**Reply by trinhpxl** [reply]
i think it might take multiple hours D: the backup on gdrive is halfway but as for the file itself, i tried 5 times and it never uploaded. will the backup version do ?

**Reply by keiwando** [reply]
Yes, the backup version will also be useful. Just let that one upload for now and I'll take a look at it and if really necessary I'll ask you for the latest version later.
Did you get some of your drawings back after importing the backup?

**Reply by trinhpxl** [reply]
backup uploaded and emailed to ya ! and i see then ! and no i only managed to import the latest version if the backup (pink square one),  i’ll check the other versions though ! 

edit : other versions wouldnt import, only the copy of the file

**Reply by keiwando** [reply]
Thank you! Yes please check the other backup as well.
Sorry about this!

**Reply by keiwando** [reply]
How big is the exported copy of the file?

**Reply by trinhpxl** [reply]
2.79 GB

**Reply by keiwando** [reply]
And how big are the backup files? 375MB?

**Reply by trinhpxl** [reply]
backup 3 is 375 mb, yes

**Reply by keiwando** [reply]
Okay, yeah that's the size of the file that you sent me. So the Google Drive upload did finish but it looks like the backup failed for some reason, since the backup file is indeed corrupted / not complete and can therefore not be loaded. I think you'll need to upload the exported copy of the project so that I can take a look.

**Reply by trinhpxl** [reply]
on it !

**Reply by keiwando** [reply]
What's the resolution and length of the video in the project?

**Reply by trinhpxl** [reply]
1920 x 1080
length of vid is 3 mins long and 2520 frames long

**Reply by trinhpxl** [reply]
it seems gdrive cannot handle the copy file, the file keeps offing itself as soon as it reaches a lil bit of the upload bar

**Reply by keiwando** [reply]
Can you try it with iCloud?
https://support.apple.com/en-gb/guide/icloud/mm708256356b/icloud

**Reply by trinhpxl** [reply]
yes ! i believe i shared it ?

**Reply by keiwando** [reply]
Yes, that worked. I could copy the file. Thank you! You can remove me from the sharing permissions again.

**Reply by trinhpxl** [reply]
you’re welcome :] and okay !

**Reply by keiwando** [reply]
Were a lot of those missing drawings unmodified duplicates or did they mostly have unique contents?

**Reply by trinhpxl** [reply]
unmodified duplicates mostly ! well, modified duplicates, only thing i changed was the foot each frame

**Reply by keiwando** [reply]
If you try to create a backup of this project, does it still fail?

**Reply by trinhpxl** [reply]
nope ! it went thru on the copy ! now i have a backup 4

**Reply by keiwando** [reply]
And the new backup is ~2.79 GB?

**Reply by trinhpxl** [reply]
same for og file, backup successful !

**Reply by trinhpxl** [reply]
> And the new backup is ~2.79 GB?

yup !

**Reply by keiwando** [reply]
By the way, the project contains a ton of drawings that are not referenced in the project itself, which is why the file is this large (it should only be around 200MB). This might have happened during the crashes. I will see if any of those drawings maybe contains the missing data so that I could send them to you manually.
I will of course also investigate how this could have happened in the first place since there are various safe-guards in place to prevent you from losing any previously saved drawings during a crash, but apparently those were not enough for some reason.

**Reply by trinhpxl** [reply]
okay, thank you so much !

**Reply by keiwando** [reply]
It's the least I can do. This should never happen!

**Reply by keiwando** [reply]
I can see 6 unique pink drawings that were apparently lost but most of the data in that project are duplicate saves of this image, which definitely appears to be a bug: 

![1DAB4F89-73A2-49BC-8FFA-E9C4B9709AD3](https://user-images.githubusercontent.com/12989570/208236538-c719d128-6886-4c5c-891e-3ba83e485d36.png)

I would recommend temporarily not working on this project until I figure out what the issue with that was, submit an update and then I can send you a fixed version of the project that is smaller (about 40MB). I'm afraid I won't be able to bring back the lost drawings unfortunately.

**Reply by keiwando** [reply]
You imported that as an image, right?

**Reply by trinhpxl** [reply]
yea i did ! and okay noted !

**Reply by keiwando** [reply]
Have you used that imported image anywhere in the animation or have you so far only imported it in preparation to later use it at some point?
I can't seem to find it used in any drawing but maybe I'm just missing it.

**Reply by keiwando** [reply]
Next follow-up. Is the original audio file that you imported called YELLOW AUDIO.wav? Is it a .wav file or does the original have a different format? And how did you import it? Via the library or the timeline? In the screenshot it looks like it loads properly for you. Is it still working?

**Reply by trinhpxl** [reply]
yea i imported it to ise it later at some point ! the audio is originally a wav ! i imported it from the timeline 
and yeah it works !

**Reply by keiwando** [reply]
If you go to "On my iPad > ToonSquid > Projects > 9C06CB39-E061-4960-90EC-207A963E655D > Audio" (in the Files app), how many files do you see there and do they all have a random-looking ID as the name or is one of them named "YELLOW AUDIO"?

**Reply by trinhpxl** [reply]
yup, all of them have random ids except ome ! 
![39426857-A17F-43F7-ACCB-F46D68374930](https://user-images.githubusercontent.com/115050959/208254677-91630e3c-2e47-442a-bdd1-fec524016117.jpeg)

**Reply by keiwando** [reply]
Thank you! Just to double-check: Did you delete the original project and then import the previously exported copy again or did you leave the original in place before importing it again and have two copies of this project now? If you have two copies, could you please check if the audio works in the other one as well?

**Reply by trinhpxl** [reply]
i left the og project in place before importing it again, yeah ! i deleted the audio from the og file bc originally the audio somehow went blank ?

**Reply by keiwando** [reply]
Okay, yeah, I'm seeing the same thing then. Thanks for all of the info, it is definitely very helpful!

**Reply by trinhpxl** [reply]
you’re welcome !

**Reply by keiwando** [reply]
After importing the image it gets automatically inserted into the timeline so you probably then deleted it. Do you happen to remember if before deleting that layer you maybe accidentally tapped on "Rasterize"?

**Reply by trinhpxl** [reply]
i did, actually !

**Reply by keiwando** [reply]
I cleaned up the project file a bit and removed all of the unnecessary leftovers of that layer rasterization and fixed the audio and sent you a link via e-mail to download it. The file is now only about 50MB large and should therefore be easier to manage. Let me know if you didn't receive the Google Drive invite. It took about seven attempts before it seemingly worked.

The layer rasterization combined with a later crash would explain the leftover data in the project without the large project size being caused by a bug. However, there are still some other issues that I'm actively investigating, e.g. why the backup failed at all, why some drawings were lost after the crash and how the audio broke in the first place. I will let you know whenever I figure out the causes of these issues and have updates that fix them.

I've released an update (1.7.2) that contains some stability fixes and fixes one very rare occurrence that could have theoretically lead to some data loss although I don't think that bug was causing the issues in this project. In any case, I recommend updating if you haven't already. 

If you ever run into any backups failing, I'd recommend contacting me if possible and also checking on the previous backup files (probably also copying them somewhere else just to be safe). I was so far unable to reproduce any backup failures with your project.

Lastly, which iPad model and iOS version are you using and do you have the following settings in the settings app enabled?
`Privacy & Security > Share iPad Analytics & Share With App Developers`
It's completely up to you whether you want to enable these or not but if you had them enabled during the crash, I might have received an anonymus report that could potentially given me a little bit of information on what might have happened.

**Reply by trinhpxl** [reply]
ipad pro 11 inch, 3rd gen !
and ios 16
i didnt have the settings u mentioned enabled at the time, but now i have incase more stuff happens !
also thank u so much :]

**Reply by keiwando** [reply]
I released the 1.7.3 update which contains some important stability fixes, especially regarding this kind of potential data loss. I found two extremely rare scenarios in which such data loss was possible, one of which was if the app crashed during a very small window of time while it was saving the project. Both of these are now fixed, so I highly recommend updating if you haven't already.
